src/cargo/core/compiler/build_runner/mod.rs

```diff
@@ -117,7 +119,11 @@ impl<'a, 'gctx> BuildRunner<'a, 'gctx> {
         };
 
         let locking_mode = if bcx.gctx.cli_unstable().build_dir_new_layout {
-            LockingMode::Fine
+            let mode = determine_locking_mode(&bcx);
+            if matches!(mode, LockingMode::Coarse) {
+                bcx.gctx.shell().warn("ulimit was to low to safely enable fine grain locking. falling back to coarse grain locking")?;
```

Comment:
This feels too disconnected from the decision that if we change it, we could easily cause a wrong message to be shown